### PR TITLE
Implemented address UI functions

### PR DESCRIPTION
**Pull request by YongchangLiu98**
We haven't implemented Delete address yet, but I tried to implement it in rest_api.js.

### CODE REVIEW

File: service/static/index.html [hunk 1]
```diff
@@ -162,7 +164,7 @@ <h5>Retrieved Address Results:</h5>
                 <th style="width:20%">Street Address</th>
                 <th style="width:15%">City</th>
                 <th style="width:15%">State</th>
-                <th style="width:15%">Zicode</th>
+                <th style="width:15%">Zipode</th>
```
Comment (ryanjcohen) [inline service/static/index.html:167]: we just have a typo in Zipcode here

File: service/static/index.html [hunk 2]
```diff
@@ -114,10 +114,12 @@ <h4>Address Information:</h4>
             <div class="form-horizontal">
               <div class="form-group">
                 <div class="col-sm-offset-2 col-sm-10">
-                  <button type="submit" class="btn btn-primary" id="addr-search-btn">Search for Customer Addresses</button>
+                  <button type="submit" class="btn btn-primary" id="addr-query-btn">List A Customer' Addresses</button>
```
Comment (ryanjcohen) [inline service/static/index.html:117]: For customers, it's setup so that list is the same as searching without any of the query parameter fields filled in. wondering if we should do the same here

File: service/static/js/rest_api.js [hunk 3]
```diff
@@ -393,5 +435,176 @@ $(function () {
...
+
+
+// ************************************************************************************************************************
+// Address Functions
+// ************************************************************************************************************************
+
+    // ****************************************
+    // Create an Address
+    // ****************************************
+    $("#addr-create-btn").click(function () {
```
Comment (ryanjcohen) [inline service/static/js/rest_api.js:447]: When I create an address using the UI, the customer's information is cleared out of the customer part of the form. Not sure if we want to show the customer information after an address is added

File: service/static/index.html (hunk 2 above)
Comment (YongchangLiu98) [reply service/static/index.html:117]: > For customers, it's setup so that list is the same as searching without any of the query parameter fields filled in. wondering if we should do the same here

Yes I just found out that I misunderstood the function I needed to implement. I will fix it.

File: service/static/js/rest_api.js (hunk 3 above)
Comment (YongchangLiu98) [reply service/static/js/rest_api.js:447]: > When I create an address using the UI, the customer's information is cleared out of the customer part of the form. Not sure if we want to show the customer information after an address is added

You mean we should let the customer's information stay? ok

File: service/static/index.html (hunk 2 above)
Comment (ryanjcohen) [inline service/static/index.html:117]: Wondering if we should change this button to Search for Customer Address and the button currently labeled as Search for Customer Address to "Retrieve Address", I think if the button labels are kept similar to the way they are for the customer portion of the UI, it makes it easier to figure out how to use

File: service/static/js/rest_api.js [hunk 4]
```diff
@@ -393,5 +435,176 @@ $(function () {
...
+        ajax.fail(function(res){
+            flash_message("Server error!")
+        });
+    });
+
+*/
+    // ****************************************
+    // //RETRIEVE A CUSTOMER'S ADDRESSES
+    // ****************************************
+    $("#addr-query-btn").click(function () {
```
Comment (ryanjcohen) [inline service/static/js/rest_api.js:581]: our api supports query parameters for addresses, and so I think it would be good to add support for that to this function

File: service/static/js/rest_api.js [hunk 5]
```diff
@@ -393,5 +435,176 @@ $(function () {
...
+            type: "GET",
+            url: "/customers/" + customer_id+"/addresses/" +address_id,
+            contentType: "application/json",
+        })
+
+        ajax.done(function(res){
+            clear_form_data()
+            add_single_address(res)
+            update_address_form_data(res)
+            flash_message("Success")
```
Comment (ryanjcohen) [inline service/static/js/rest_api.js:541]: I think it might be easier user-wise if the retrieved address is shown on the actual form instead of the retrieved address results. because the retrieved address results isn't visible on my browser until i scroll down, it almost looks like this button doesn't work until I scroll down

File: service/static/js/rest_api.js (hunk 4 above)
Comment (ryanjcohen) [reply service/static/js/rest_api.js:581]: i also think that within the form, we should make sure that the customer id is not erased, and can also have the first address in the customer's list of address show up in the form (that way it's similar to how "Search for Customer" results show

**Comment by ryanjcohen** [comment]
one general thought is that i think we need to add in clearing out of the retrieved customer results and retrieved address results when a new request is made and the data they're showing is no longer relevant. we could do this in a new issue though

File: service/static/index.html (hunk 2 above)
Comment (YongchangLiu98) [reply service/static/index.html:117]: no problem

File: service/static/js/rest_api.js (hunk 4 above)
Comment (YongchangLiu98) [reply service/static/js/rest_api.js:581]: Yes that's where I misunderstood the function, I will fix them.

File: service/static/js/rest_api.js (hunk 5 above)
Comment (YongchangLiu98) [reply service/static/js/rest_api.js:541]: > I think it might be easier user-wise if the retrieved address is shown on the actual form instead of the retrieved address results. because the retrieved address results isn't visible on my browser until i scroll down, it almost looks like this button doesn't work until I scroll down

Makes sense.

**Comment by ryanjcohen** [comment]
> one general thought is that i think we need to add in clearing out of the retrieved customer results and retrieved address results when a new request is made and the data they're showing is no longer relevant. we could do this in a new issue though

@YongchangLiu98 line 48 of https://github.com/devops-customers-squad/customers/blob/main/service/static/js/rest_api.js shows how I've done this for customers, I think we should consider doing the same for addresses

File: service/static/index.html (hunk 1 above)
Comment (ryanjcohen) [inline service/static/index.html:167]: We still have a typo here we need to fix, it currently reads "Zipode"

File: service/static/js/rest_api.js [hunk 6]
```diff
@@ -19,17 +19,13 @@ $(function () {
     }
 
     function update_address_form_data(res) {
-        if (res != null && res.addresses.length != 0) {
-            var address = res.addresses[0];
```
Comment (ryanjcohen) [inline service/static/js/rest_api.js:23]: This change is introducing a bug for customer specific functions (see line 326). Is it possible to undo the changes here and create a separate function for updating the address form given the response from an address subordinate api endpoint

File: service/static/js/rest_api.js [hunk 7]
```diff
@@ -30,6 +30,14 @@ $(function () {
         }
     }
 
+    function update_address_form(res) {
```
Comment (ryanjcohen) [inline service/static/js/rest_api.js:33]: this throws an error in the browser when res is null, maybe just add a conditional check to make sure that res is not null before trying to access res.address_id

File: service/static/js/rest_api.js [hunk 8]
```diff
@@ -48,6 +56,16 @@ $(function () {
         $("#customer_table").append(create_customer_results_header());
```
Comment (ryanjcohen) [inline service/static/js/rest_api.js:56]: I think we need to add something similar here for the address_table, otherwise I'm not sure that we risk that bdd tests could fail in some cases because past search results haven't been cleared out

**Review by ryanjcohen — APPROVED** [review]
